pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go

```diff
@@ -360,33 +361,72 @@ func (ca *cloudCIDRAllocator) updateCIDRAllocation(nodeName string) error {
-				break
-			}
-		}
+		node.Spec.PodCIDR = cidrStrings[0]
+		node.Spec.PodCIDRs = cidrStrings
 	}
+
+	err = ca.updateNodeCIDR(node, oldNode)
 	if err != nil {
-		nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRAssignmentFailed")
-		klog.ErrorS(err, "Failed to update the node PodCIDR after multiple attempts", "nodeName", node.Name, "cidrStrings", cidrStrings)
 		return err
 	}
-
-	if northInterfaces != nil || additionalNodeNetworks != nil {
-		if err := ca.updateMultiNetworkAnnotations(node, northInterfaces, additionalNodeNetworks); err != nil {
+	if !reflect.DeepEqual(node.Annotations, oldNode.Annotations) {
+		if err = utilnode.PatchNodeMultiNetwork(ca.client, node); err != nil {
+			nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRAssignmentFailed")
+			klog.ErrorS(err, "Failed to update the node annotations and capacity for multi-networking", "nodeName", node.Name)
 			return err
 		}
 	}
-	err = utilnode.SetNodeCondition(ca.client, types.NodeName(node.Name), v1.NodeCondition{
-		Type:               v1.NodeNetworkUnavailable,
-		Status:             v1.ConditionFalse,
-		Reason:             "RouteCreated",
-		Message:            "NodeController create implicit route",
-		LastTransitionTime: metav1.Now(),
+	return err
+}
+
+func (ca *cloudCIDRAllocator) setNetworkCondition(node *v1.Node) {
+	for i := range node.Status.Conditions {
+		if node.Status.Conditions[i].Type == v1.NodeNetworkUnavailable {
+			// we do not update Times so that we do not trigger unnecessary updates
+			node.Status.Conditions[i].Status = v1.ConditionFalse
+			node.Status.Conditions[i].Reason = "RouteCreated"
+			node.Status.Conditions[i].Message = "NodeController create implicit route"
+			return
+		}
+	}
+	// NodeNetworkUnavailable condition not found, lets add it to trigger update,
+	// proper values are set during Patch.
+	node.Status.Conditions = append(node.Status.Conditions, v1.NodeCondition{
+		Type: v1.NodeNetworkUnavailable,
 	})
-	if err != nil {
-		klog.ErrorS(err, "Error setting route status for the node", "nodeName", node.Name)
+}
+
+func (ca *cloudCIDRAllocator) updateNodeCIDR(node, oldNode *v1.Node) error {
+	var err error
+
+	// update Spec.podCIDR
+	if !reflect.DeepEqual(node.Spec, oldNode.Spec) {
+		// TODO: remove the retry since it is handled by the reconciliation loop
+		for i := 0; i < cidrUpdateRetries; i++ {
+			if err = utilnode.PatchNodeCIDRs(ca.client, types.NodeName(node.Name), node.Spec.PodCIDRs); err == nil {
+				klog.InfoS("Set the node PodCIDRs", "nodeName", node.Name, "cidrStrings", node.Spec.PodCIDRs)
+				break
+			}
+		}
+		if err != nil {
+			nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRAssignmentFailed")
+			klog.ErrorS(err, "Failed to update the node PodCIDR after multiple attempts", "nodeName", node.Name, "cidrStrings", node.Spec.PodCIDRs)
+			return err
+		}
+	}
+
+	// Update Conditions
+	if !reflect.DeepEqual(node.Status.Conditions, oldNode.Status.Conditions) {
```

Comment:
you already set this condition here https://github.com/kubernetes/cloud-provider-gcp/pull/512/commits/14ad25f8ae106a4d2ddd00b87fe7a7b6a03c696d?diff=split&w=0#r1184318055, why set it here too?

Reply:
also, this is `updateNodeCIDR` , I think that you only have to update the Condition at the end, to indicate that the node is ready

Reply:
updating Node's CIDR comes with setting NodeNetworkUnavailable condition. let me just set that value in setNetworkCondition, and then get that condition here, so we only define it once.